### ISSUE TITLE
Code splitting and i18n files

### ISSUE THREAD

**Issue by slorber** [gatsbyjs/gatsby#13806]
Hi,

The focus of this issue/question is not related to making i18n work, but is related to making it fast (ie, make code splitting work in an optimized way for your translation bundles).

There are already [issues](https://github.com/gatsbyjs/gatsby/issues/3853) regarding how to translate your Gatsby website with various tools/libs/CMS. There are also some nice blog posts on how to create one page per lang (hint, loop over your locales and call `createPages`, or read [this post](https://blog.significa.pt/i18n-with-gatsby-528607b4da81)

 But so far, all the solutions mentioned focus on making it work, but they make it work in an unoptimized way.

Here I'll focus mostly on splitting per language. If you have to split translations by page, the best of probably to use a CMS and inject page specific data from context/query. What I focus on is mostly translations that are "global" into your app, like the translations appearing in your main layout that you reuse everywhere.

# Bad solution 1: importing all locale files in your layout

If you do this in your layout component: 

```
import enData from 'react-intl/locale-data/en'
import ptData from 'react-intl/locale-data/pt'
import en from '../i18n/en.json'
import pt from '../i18n/pt.json'
```

Then, all the translations will always be downloaded by your user inside the js bundle, for any page of your site. 

This means that if you have 50 languages and a very large static site, this solution will download a lot of useless bytes. The more your static site grows, the more there is a delay between the static page being rendered and the static page being interactive (ie you can't click on the buttons yet).

This is to me a big problem and makes this solution unsuitable

# Bad solution 2: providing locale messages with pageContext

If you do this in `gatsby-node.js`

```js
["fr","en"].map(locale => {
  createPage({
    path: `/${locale}/${page.path}`
    context: {
      messages: getMessagesForLocale(locale);
    }
  })
})
```

Then you can use the messages of your locale from the context.

The problem is that the french locale messages will be in the context of every french page. It means, everytime you navigate to another french page, you will redownload everytime those translations as it's included into all pages js bundles.

# Possible solution

@KyleAMathews mentioned [here](https://github.com/gatsbyjs/gatsby/issues/3853#issuecomment-367115380) that we could:
- Create on layout per lang
- Create on "wrapper component" per page, per locale

I can see how this would solve the code splitting issues, and how webpack would be able to move some translation files into "common chunks", but this is a dev maintainance burden not really sustainable.

This means that for 2 locales and 2 pages you would have:
- EnLayout
- FrLayout
- EnPage1
- FrPage1
- EnPage2
- FrPage2

Now for 3 locales and 3 pages you would have:
- EnLayout
- FrLayout
- DeLayout
- EnPage1
- FrPage1
- DePage1
- EnPage2
- FrPage2
- DePage2
- EnPage3
- FrPage3
- DePage3

Now, imagine that you have 10 pages and 10 locales (which I would not consider a large international website), you end up with more than 1000 entry point files to maintain, most of them look almost similar.

# Possible solution: using the layout plugin

This [comment](https://github.com/gatsbyjs/gatsby/issues/3830#issuecomment-363326165) of @KyleAMathews  (and some other parts of the thread), is helpful and 

However, I'm not even sure this solution works in v2, because the [plugin doc](https://www.gatsbyjs.org/packages/gatsby-plugin-layout/) that not mention we can use different layouts in `createPages` while according to [this comment](https://github.com/gatsbyjs/gatsby/issues/3830#issuecomment-362710469) it seems v1 allowed to do this:

```js
posts.edges.map(({ node }) => {
  const id = node.contentfulid;
  const locale = node.node_locale;

  return createPage({
    path: `/${locale}/blog/posts/${id}`,
    layout: locale,
    component: path.resolve('./src/templates/post-page.jsx'),
    context: {
      id,
      locale,
    },
  });
});
```

Is it still possible to use different layout (in term of layout plugin) from createPages?

# What should we do?

I'd like to know what is the current recommended way to solve this problem, that is suitable for a large but realistic website (something like 50 pages and 20 locales) and optimized.

cc @KyleAMathews @deltaskelta @szimek @pbrandone

**Comment by universse** [comment]
It seems https://github.com/kentcdodds/babel-plugin-preval / https://github.com/kentcdodds/preval.macro is a nice solution to this. preval.macro works out of the box with gatsby. See https://www.gatsbyjs.org/docs/babel-plugin-macros/

First during createPages, you inject locale into page context.
```javascript
posts.edges.map(({ node }) => {
  const id = node.contentfulid;
  const locale = node.node_locale;

  return createPage({
    path: `/${locale}/blog/posts/${id}`,
    component: path.resolve('./src/templates/post-page.jsx'),
    context: {
      id,
      locale,
    },
  });
});
```

Then, in your layouts/index.js (I'm assuming you are using gatsby-plugin-layout)
```javascript
import preval from 'preval.macro'

export default function Layout ({ pageContext: { locale } }) {
  // with preval.macro you can use node apis (like fs.readFile) and have the data at build time
  const translations = preval`
    const fs = require('fs')
    // here interpolate pageContext.locale to to read the correct translation file.
    const translations = fs.readFileSync('${locale}.json')
    module.exports = translations
  `
  
  // LanguageProvider use Context api
  return ( 
    <LanguageProvider value={translations}>
      {children}
    </LanguageProvider>
}
```
This way you only need 1 layout file instead of 1 for each language.

One last optimization. Since translations is a new object every render, it can be memoized with useMemo hook.

**Comment by slorber** [comment]
That looks like a great solution if this works :)  

I thought of using macros but wasn't sure it would work because I thought macros would not be able to get access to the locale variable but I may be wrong.

Have you tried this approach and confirm it works or it's an idea to test?

**Comment by universse** [comment]
Hm. Ok last time I made use of this technique, I didn't interpolate variables from react props. Thought it could work here also. Take a look here https://codesandbox.io/s/6l7n7pvkr3

**Comment by slorber** [comment]
Yeah this works because the locale is hardcoded/static, but don't think it's suitable to solve the problem.

**Comment by szimek** [comment]
I'd love to see an official solution to this issue.

After migrating from v1 to v2, we've started to build the app separately for each language as a workaround for this issue, and some other ones, like huge size of `pages-manifest.js` file and lack of `assetPrefix` option back then.

**Comment by slorber** [comment]
Hi and thanks for the feedback @szimek . I did not talk about your growing pages manifest but yes that looks like a potential issue for scalability I didn't think about.

Even after reading your comment [here](https://github.com/gatsbyjs/gatsby/issues/3830#issuecomment-491505582) I'm not totally sure what you mean by using asset prefix? Do you create one prefix per locale? What problem does it solve for you?

**Comment by slorber** [comment]
The kind of API I would like in Gatsby would enable this.

```js

const AppLocaleList = ["fr_FR","en-US"];
const TranslatedPages = [
    {
      path: '/items',
      component: path.resolve(`src/page/items.tsx`),
    },
    {
      path: '/item/:itemId',
      matchPath: '/item/:itemId',
      component: path.resolve(`src/page/item.tsx`),
    },
];

TranslatedPages.forEach(translatedPage => {
    AppLocaleList.forEach(locale => {
      const prefix = getLocalePrefix(locale);
      const messages = AppLocalesMap[locale].messages;
      const page = {
        path: prefix + translatedPage.path,
        component: [
          path.resolve(`src/layouts/${locale}_layout.tsx`)
          translatedPage.component,
        ],
      };
      if (translatedPage.matchPath) {
        page.matchPath = prefix + translatedPage.matchPath;
      }
      createPage(page);
    });
  });
```


The ability to pass a component array would be convenient to create multiple "wrappers", applied in order, which all have the right imports. This would enable to optimize bundles on multiple axes and be flexible enough for multiple further usecases.

**Comment by tricoder42** [comment]
Hey everyone, I'm working on automatic splitting of message files in https://github.com/lingui/js-lingui/issues/503. Ideally I would like to have a webpack plugin, which:

1. Generates message files for each chunk, e.g. `chunk.0.[locale].js` (I've already done this, because I can track what message is included in what module)
2. When `chunk.0.js` is about to load, request `chunk.0.[locale].js` at the same time. That requires runtime function, which decide what locale to load. It also requires Webpack Plugin API available in Webpack 5. @sokra from webpack pointed me to [LocalizationPlugin](https://github.com/webpack/webpack/blob/next/test/configCases/custom-source-type/localization/webpack.config.js) example which uses new API.

I want also experiment with [per-locale bundles](https://github.com/lingui/js-lingui/issues/508). That means that instead of having code, which loads translations at runtime, each chunk is generated for every locale with translations baked in. I believe this would work great with static sites served under `[locale]/` url prefix and it's actually _easy_ to do with new version of LinguiJS.

That said, both solutions are waiting for next version of some libraries, but it would be great to get a feedback or help from you guys.

**Comment by danilobuerger** [comment]
@slorber did you make any progress on finding a good solution?

**Comment by slorber** [comment]
Hi @danilobuerger unfortunately I've not yet solved this problem and currently not working on this anymore. 

In my understanding, it will be better to pass heavy text content directly through context (probably easier with a cms integration, as you can assign a translation directly to a specific page).

Yet I'm still interested to a solution to this problem, as having local non-cms keys still make sense in a few usecases (for translating your global site layout for example)


I've seen a blog post published recently (still in my reading list), maybe it will give some infos on this topic, or its author @ondrabus has an opinion? 
https://www.gatsbyjs.org/blog/2020-02-19-how-to-build-multilingual-sites-with-gatsby/

**Comment by danilobuerger** [comment]
Alright. I am currently experimenting with i18next (but this could be any framework) and lazy json loading. It seems that those will be code splitted separately and loaded only once (not again on page nav, like the page context option would) and only the current locale (unlike the import all solution at the top). I will report back (if I don't forget) once I manage to get it into production!

**Comment by danilobuerger** [comment]
I think there a two acceptable solutions on where the translation data should end up:

* Persist only the translation strings which the current page / locale needs in the page context. This is a variation of your outlined `Bad solution 2`, but only persists what is really needed. This would be equivalent to how the remark plugin usually works, persisting the currently needed markdown (and only that) to the page data. This approach would make sense if you have lots of pages with different content.

* Persist each locales translation strings to a separate file (this could be n files, arbitrarily splitting the translations, for example by namespaces) and load that file along with the page data. When navigating (and keeping the locale the same) this file would not be requested again. This approach would make sense if you have lots of common translation strings.

Since we have SSR and the first page would already be localized by it, the translation file could be loaded after the page load but before navigating. This could be quicker than having a subset of translation strings in the page data. So I am going to investigate the second option.

The approach could be as follows:
1. SSR:
  * Render the page in the current locale
  * Collect the used translation strings
  * Render the collected translation strings into the body as a script tag to make them available for hydration

2. Browser:
  * Hydrate the translation strings from a special `window` var. This will prevent a content flash, as the browser rendering will take over but didn't load the translation file yet
  * Use webpacks lazy import to request the whole translation file

3. Changing Locale:
  * Use suspend to wait while loading the new translation file

**Comment by slorber** [comment]
Thanks @danilobuerger that seems a cool idea, I didn't think about it but it could work. Somehow it's the same strategy that is applied to extract the critical CSS of css-in-js libs, but instead of css, it's translation strings

**Comment by LekoArts** [comment]
cc @pieh - we wanted to maybe work on a PoC here

**Comment by ddennis** [comment]
I found [https://dev.to/jospohr/proper-i18n-with-gatsby-i18next-and-sanity-io-5g8a](https://dev.to/jospohr/proper-i18n-with-gatsby-i18next-and-sanity-io-5g8a ) which seems to solve most of the problems with "bad solution 1" and "bad solution 2". repo [here](https://github.com/johannesspohr/gatsby-starter-i18next-sanity)

Does anybody have insights to why this should not be the way to go? 
I would assume the sourcing from Sanity can just be replaced with whatever alternativ if needed.

Ny feedback would be much appreciated.

**Comment by LekoArts** [comment]
The person is still passing the data via `pageContext`, or did I miss something? (Which is "Bad solution 2")

**Comment by ddennis** [comment]
> The person is still passing the data via `pageContext`, or did I miss something? (Which is "Bad solution 2")

Yes you are right. But only the translation keys for the specific/requested page.
but is it not the solution as described by **danilobuerger**  

> Persist each locales translation strings to a separate file (this could be n files, arbitrarily splitting the translations, for example by namespaces) and load that file along with the page data. When navigating (and keeping the locale the same) this file would not be requested again. This approach would make sense if you have lots of common translation strings.

The whole i18Next instance is in the pageContext.

**Comment by ddennis** [comment]
@LekoArts i saw you have been working on **https://github.com/LekoArts/gatsby-theme-i18n** implementing examples - would it make sense to start using your solution? or should we wait for a more official release?

**Comment by LekoArts** [comment]
This is currently WIP and not released yet.

**Comment by nsavareika** [comment]
> I think there a two acceptable solutions on where the translation data should end up:
> 
> * Persist only the translation strings which the current page / locale needs in the page context. This is a variation of your outlined `Bad solution 2`, but only persists what is really needed. This would be equivalent to how the remark plugin usually works, persisting the currently needed markdown (and only that) to the page data. This approach would make sense if you have lots of pages with different content.
> * Persist each locales translation strings to a separate file (this could be n files, arbitrarily splitting the translations, for example by namespaces) and load that file along with the page data. When navigating (and keeping the locale the same) this file would not be requested again. This approach would make sense if you have lots of common translation strings.
> 
> Since we have SSR and the first page would already be localized by it, the translation file could be loaded after the page load but before navigating. This could be quicker than having a subset of translation strings in the page data. So I am going to investigate the second option.
> 
> The approach could be as follows:
> 
> 1. SSR:
> 
> * Render the page in the current locale
> * Collect the used translation strings
> * Render the collected translation strings into the body as a script tag to make them available for hydration
> 
> 1. Browser:
> 
> * Hydrate the translation strings from a special `window` var. This will prevent a content flash, as the browser rendering will take over but didn't load the translation file yet
> * Use webpacks lazy import to request the whole translation file
> 
> 1. Changing Locale:
> 
> * Use suspend to wait while loading the new translation file

Another possible solution to prevent a content flash on initial load is to perform hydration only after translation file has loaded. Gatsby browser api provides [replaceHydrateFunction](https://www.gatsbyjs.com/docs/browser-apis/#replaceHydrateFunction) which allows replacing `ReactDOM.hydrate`/ `ReactDOM.render` function with the custom one. You could use it to load translation file depending on the requested locale, place translations inside `window` var, and only then perform the rendering. Then you won’t have to store script with translations on every generated html page. Example:
```
exports.replaceHydrateFunction = () => {
	return (element, container, callback) => {
		const language = getLanguageFromPath(window.location.pathname);

		import(`./src/translations/${language}.json`).then(translations => {
                      window.__translations = translations.default;
                      ReactDOM.render(element, container, callback);
                });
	};
};
```
